File: haskell/ch1/Ex110.hs
module Ex110 where

a _ 0 = 0
a _ 1 = 2
a x y = if x==0 then 2 * y else a (pred x) (a x (pred y))

f n = a 0 n
g n = a 1 n
h n = a 2 n
k n = 5 * n * n

File: elixir/ch1/ex110.ex
defmodule Ex110 do
  @moduledoc false

  def ack(_, 0), do: 0
  def ack(0, y), do: 2 * y
  def ack(_, 1), do: 2
  def ack(x, y), do: ack(x - 1, ack(x, y - 1))

  def f(n), do: ack(0, n)

  def g(n), do: ack(1, n)

  def h(n), do: ack(2, n)

  def k(n), do: 5 * n * n

end
